govcd/user.go

```diff
@@ -0,0 +1,505 @@
+/*
+ * Copyright 2019 VMware, Inc.  All rights reserved.  Licensed under the Apache v2 License.
+ */
+
+package govcd
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/vmware/go-vcloud-director/v2/types/v56"
+	"github.com/vmware/go-vcloud-director/v2/util"
+)
+
+// Definition of an OrgUser
+type OrgUser struct {
+	User     *types.User
+	client   *Client
+	AdminOrg *AdminOrg // needed to be able to update, as the list of roles is found in the Org
+}
+
+// Simplified structure to insert or modify an organization user
+type OrgUserConfiguration struct {
+	Name            string // Mandatory
+	Password        string // Mandatory
+	RoleName        string // Mandatory
+	ProviderType    string // Optional: defaults to "INTEGRATED"
+	IsEnabled       bool   // Optional: defaults to false
+	IsLocked        bool   // Only used for updates
+	DeployedVmQuota int    // Optional: 0 means "unlimited"
+	StoredVmQuota   int    // Optional: 0 means "unlimited"
+	FullName        string // Optional
+	Description     string // Optional
+	EmailAddress    string // Optional
+	Telephone       string // Optional
+}
+
+const (
+	// Common role names and provider types are kept here to reduce hard-coded text and prevent mistakes
+	// Roles that are added to the organization need to be entered as free text
+
+	OrgUserRoleOrganizationAdministrator = "Organization Administrator"
+	OrgUserRoleCatalogAuthor             = "Catalog Author"
+	OrgUserRoleVappAuthor                = "vApp Author"
+	OrgUserRoleVappUser                  = "vApp User"
+	OrgUserRoleConsoleAccessOnly         = "Console Access Only"
+	OrgUserRoleDeferToIdentityProvider   = "Defer to Identity Provider"
+
+	// Allowed values for provider types
+	OrgUserProviderIntegrated = "INTEGRATED" // The user is created locally or imported from LDAP
+	OrgUserProviderSAML       = "SAML"       // The user is imported from a SAML identity provider.
+	OrgUserProviderOAUTH      = "OAUTH"      // The user is imported from an OAUTH identity provider
+)
+
+// Used to check the validity of provider type on creation
+var OrgUserProviderTypes = []string{
+	OrgUserProviderIntegrated,
+	OrgUserProviderSAML,
+	OrgUserProviderOAUTH,
+}
+
+// NewUser creates an empty user
+func NewUser(cli *Client, org *AdminOrg) *OrgUser {
+	return &OrgUser{
+		User:     new(types.User),
+		client:   cli,
+		AdminOrg: org,
+	}
+}
+
+// Private function to perform the user search by either name or ID
+// It will only look for one of the two fields, depending on the value of byName
+func (adminOrg *AdminOrg) fetchUserByNameOrId(identifier string, byName, refresh bool) (*OrgUser, error) {
```

Comment:
Please split this function into two separate ones as it would aid code readability a lot. Now places like these require comments:

```
        // First look by ID
	orgUser, err := adminOrg.fetchUserByNameOrId(identifier, false, true)
	// if it fails, look by name
	if IsNotFound(err) {
		orgUser, err = adminOrg.fetchUserByNameOrId(identifier, true, false)
	}
```

But would become self-explanatory:

```
        orgUser, err := adminOrg.fetchUserByName(identifier, true)
        if IsNotFound(err) {
            orgUser, err = adminOrg.fetchUserById(identifier, false)
        }
```

Even these new functions would be easier to read as would avoid error-prone ifs like these:

```
		if !byName && orgUser.ID == identifier {
			userHREF = orgUser.HREF
			break
		}
		if byName && orgUser.Name == identifier {
			userHREF = orgUser.HREF
			break
		}
```

Even though it would result in a bit of code duplication, it's really tiny compared to the readability improvement.

Reply:
done

Reply:
Thanks, it's nice how you solved it with the new `FetchUserByHref(href string)` function - it may prove useful for other use cases as well!